src/v/iceberg/rest_client/tests/parser_tests.cc

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "iceberg/rest_client/parsers.h"
+
+#include <gtest/gtest.h>
+
+namespace r = iceberg::rest_client;
+
+json::Document make_doc(ss::sstring s) {
+    json::Document d;
+    d.Parse(s);
+    return d;
+}
+
+TEST(oauth_token, invalid_doc_fails) {
+    {
+        auto result = r::parse_oauth_token(make_doc("{}"));
+        ASSERT_FALSE(result.has_value());
+        auto error = result.error();
+        ASSERT_TRUE(std::holds_alternative<r::json_parse_error>(error));
+
+        const auto& parse_error = std::get<r::json_parse_error>(error);
+        ASSERT_EQ(
+          parse_error.context,
+          "parse_oauth_token::response_does_not_match_schema");
+    }
+
+    {
+        auto result = r::parse_oauth_token(
+          make_doc(R"J({"access_token": "", "expires_in": "42"})J"));
+        ASSERT_FALSE(result.has_value());
+        auto error = result.error();
+        ASSERT_TRUE(std::holds_alternative<r::json_parse_error>(error));
```

Comment:
It's possible I don't think it's really much simpler in this case but it's possible to use Matches to get better error messages and easier to read assertions. We don't have any custom matchers for `tl::expected` so we can't make this a oneliner without effort to create custom matchers, but if we're going to use `tl::expected` more it's worth doing in my opinion. In the sake of time you don't have to change anything in this PR, but good to know for future development: http://google.github.io/googletest/reference/matchers.html

Reply:
It looks like I can cut down some of the noise with `VariantWith` etc, eg

```
ASSERT_THAT(
      result.error(),
      VariantWith<json_parse_error>(Field(
        &json_parse_error::context,
        "parse_oauth_token::response_does_not_match_schema")));
```

Reply:
I've used this same pattern in a couple of other places (nested variants)